pbs_cgroup hook is getting deleted in revert_to_defaults( )

shilpakodli opened the pull request:
#### Bug/feature Description
* After PBS Installation, we have pbs_cgroups as default hook. 
     PTL deletes all site hooks in revert_to_default function including default pbs_cgroups hook

#### Affected Platform(s)
* All 

#### Solution Description
* Exclude pbs_cgroups hook while deleting site hooks.

#### Testing logs/output
*  [TestRevertSiteHooks_logs.txt](https://github.com/PBSPro/pbspro/files/3033841/TestRevertSiteHooks_logs.txt)

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
- [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
- [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

prakashcv13 commented on test/fw/ptl/lib/pbs_testlib.py:
```diff
@@ -5108,6 +5109,9 @@ def revert_to_defaults(self, reverthooks=True, revertqueues=True,
             hooks = [h['id'] for h in hooks]
+            for h in skip_site_hooks:
+                hooks.remove(h)
+            hooks.remove('pbs_cgroups')
```
I think pbs_cgroups got removed in the for loop, this is why travis is failing.
However, isn't this PR about not deleting the cgroup hook?

shilpakodli replied:
Hi Prakash, I am deleting pbs_cgroups from hooks list which will be deleted in setUp.
Travis was failing due to extra line of code which I have removed now.

hirenvadalia replied:
@shilpakodli before calling `hooks.remove('pbs_cgroups')` you need to check whether hooks has 'pbs_cgroups' in it or not else it will throw `ValueError: list.remove(x): x not in list` error

prakashcv13 replied:
I missed seeing that you are removing it from the list being sent for deletion.

prakashcv13 replied:
@hirenvadalia - I also thought of that but assumed that pbs_cgroups will always be there. But, it is better to be safe than sorry.

prakashcv13 commented on test/tests/selftest/pbs_test_revert_site_hooks.py:
```diff
@@ -0,0 +1,60 @@
+        """
+        rc = self.server.manager(MGR_CMD_CREATE, HOOK, None, "exechook")
+        self.assertEqual(rc, 0)
+        TestSelf.setUp(self)
```
We need not to create a new hook to test if cgroups hooks exists or not.
Also, I think we can remove the call to setUp() as well.

prakashcv13 commented on test/tests/selftest/pbs_test_revert_site_hooks.py:
```diff
@@ -0,0 +1,60 @@
+        else:
+            for h in hooks:
+                if h not in ['pbs_cgroups']:
+                    self.fail("Default cgroups hook doesn't exist")
```
you have already confirmed that there is only one item in hooks, so the for loop is not required.
However, there is no guarantee that we would always have only cgroups enabled by default, if we enable another hook the test for len(hooks) to be 1 will fail. So, I suggest that you just check for existence of pbs_cgroups in "hooks" list.

prakashcv13 commented on test/tests/selftest/pbs_test_revert_site_hooks.py:
```diff
@@ -0,0 +1,60 @@
+
+class TestRevertSiteHook(TestSelf):
+    """
+    Contains test to delete hooks in Setup() except cgroups hook
```
change setUp() to revert_to_defaults()

shilpakodli commented:
@prakashcv13 and @hirenvadalia Addressed your review comments and attaching latest execution logs.
[TestRevertSiteHooks_logs.txt](https://github.com/PBSPro/pbspro/files/3053517/TestRevertSiteHooks_logs.txt)

prakashcv13 commented on test/tests/selftest/pbs_test_revert_site_hooks.py:
```diff
@@ -0,0 +1,54 @@
+        """
+        Test if cgroups hook exists at the end of the test case
+        """
+        self.server.revert_to_defaults(self)
```
Do we need to call revert_to_defaults() explicitly?

prakashcv13 commented on test/tests/selftest/pbs_test_revert_site_hooks.py:
```diff
@@ -0,0 +1,54 @@
+        """
+        self.server.revert_to_defaults(self)
+        hooks = [h['id'] for h in self.server.status(HOOK)]
+        for h in hooks:
```
If in future we have more than one default hook enabled, this test will fail.
we do not need a for loop here - 
just check for the existence of 'pbs_cgroup' in hooks with if 'pbs_cgroups' in hooks

prakashcv13 commented on test/tests/selftest/pbs_test_revert_site_hooks.py:
```diff
@@ -0,0 +1,54 @@
+from tests.selftest import *
+
+
+class TestRevertSiteHook(TestSelf):
```
would TestCGroupsHookPresence() make more sense?

shilpakodli replied:
I have made this in general so that in future if we had to add hooks to skip list then this test suite will help in writing similar test cases.

shilpakodli commented:
@prakashcv13 I have addressed your comments. Please let me know if any changes needs to be done.

shilpakodli commented:
I have rebased and squashed my commits. Please take my PR forward.
